### ISSUE TITLE
ROS2 Nodes stay paused when stepping Gazebo simulation 

### ISSUE THREAD

**Issue by shonigmann** [ros-simulation/gazebo_ros_pkgs#1324]
We've noticed that when we have ROS2 nodes using `sim_time` interfacing with our Gazebo simulation, all of our nodes are unresponsive when stepping through the simulation.

The nodes behave completely as expected when running the simulation normally (unpaused). They "pause" as expected if we pause the simulation. But when we try and step the simulation, nothing happens. In Gazebo, we can see the simulation time increasing, as well as any dynamics within the simulation working as expected. But everything on the ROS front remains "paused". This is perhaps best illustrated (and perhaps caused) by the `/clock` topic. We can monitor the `/clock` topic, and confirm that it updates as expected when the simulation plays, but when we step the simulation manually, nothing is ever published. We can manually step 10 seconds ahead, and then if we unpause the simulation, the `/clock` starts publishing again with a 10 second jump. 

This is an issue for us as we'd like to have a gazebo-simulink co-simulation which works by stepping the gazebo sim for every time step in the simulink sim. But this behavior prevents anything else from working.

We've tested this in ROS2 foxy. Interestingly, in ROS1 Melodic, topics seem to continue to update if we manually step the simulation. So perhaps an implementation was changed or a feature has yet to be ported between the ROS1 and ROS2 versions of gazebo_ros_pkgs? 

Happy to help try and fix this, but not sure where to start looking. Any help would be welcome!

**Comment by alikureishy** [comment]
I had opened an issue in April (issue #1268), and had proposed a workaround (#1280) for the underlying inability to step Gazebo-Ros **together** (as a feature), but it didn't catch much attention at the time. Hopefully we can get a broader discussion about this going.

**Comment by shonigmann** [comment]
Yes, I read your issue, and thank you for your workaround. And I agree it would be great for this issue to get a bit more attention.  

Unfortunately, I think my use-case needs the inverse of what you've proposed - step control of ROS through Gazebo, as the Simulink-Gazebo plugin we are working with steps Gazebo directly. I suppose I could alter that plugin as well to instead step the simulation through a ROS service interface, but that feels a bit "clunky" to me... I'd much prefer being able to control time from the simulation environment, and have ROS behave accordingly.

**Comment by alikureishy** [comment]
Ah, I see what you're saying, @shonigmann .

Is there a step size (in ms) that the Simulink plugin uses to advance Gazebo with? If so, you might want to check what that size is. Based on what you are describing, it seems that that individual step size might not be large enough to trigger the gazebo_ros plugin to publish to /clock, even when you manually step it a few times.

**Alternatively**, on the gazebo_ros side, there is a *pub_clock_frequency* param for gazebo_ros that you could also tweak, so that the /clock publish threshold is met even with the manual step size being used by the Simulink plugin.

To illustrate more, here's the check that gazebo_ros does before publishing to /clock:
```
...
  if (pub_clock_frequency_ > 0 && (sim_time - last_pub_clock_time_).Double() < 1.0/pub_clock_frequency_)
    return;
...
```

Essentially, if that check succeeds, gazebo_ros **skips** publishing to the /clock topic. Could you check what value 'pub_clock_frequency' has at your end? You can check that by issuing this command in a ROS-sourced shell:
```
> rosparam get /gazebo/pub_clock_frequency
```

My guess is that it is set to something > 0, and is probably too low. If I'm right, you might need to either increase that value to see more updates going to the /clock topic (based on the step size being used by the Simulink plugin), or just set it to 0 (or remove that parameter entirely), so it publishes every time.

Hope that helps...

**Comment by shonigmann** [comment]
After a bit more digging, I think the issues we have been experiencing are due to the [Throttler checking the `realTime` clock instead of the `simTime` clock](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/c1793e17069b9785568956f7cb10b9fb646abed1/gazebo_ros/src/gazebo_ros_init.cpp#L260-L262)

```
void GazeboRosInitPrivate::PublishSimTime(const gazebo::common::UpdateInfo & _info)
{
  if (!throttler_.IsReady(_info.realTime)) {  // <-- HERE
    return;
  }

  rosgraph_msgs::msg::Clock clock;
  clock.clock = gazebo_ros::Convert<builtin_interfaces::msg::Time>(_info.simTime);
  clock_pub_->publish(clock);
```

For me, this begs two questions:
1) Why is realTime being used? 
2) And why does stepping the simulation not change the realTime clock?
